### DISCUSSION TITLE
USB CDC Issues on RP2350 ( Raspberry Pico forums )

### DISCUSSION THREAD

**Discussion by Serialcomms** [tabemann/zeptoforth#158] [General]
Hello All,

I've started to look at the USB/CDC issue reported on the [Raspberry Pico forum](https://forums.raspberrypi.com/viewtopic.php?p=2263566&hilit=tabemann#p2263566)

I reckon there are some USB implementation issues which I may be able to help with, based on my [USB CDC Serial Stack](https://github.com/Serialcomms/Raspberry-Pico-USB-Serial_BETA) for the RP2040.

Is anyone able to assist with the Zeptoforth-specific elements of this - i.e. modules and the console ?

**Comment by tabemann** [comment]
On Mon, Nov 25, 2024 at 10:04 AM serialcomms ***@***.***>
wrote:

> Hello All,
>
> I've started to look at the USB/CDC issue reported on the Rapsberry Pico
> forum
> <https://forums.raspberrypi.com/viewtopic.php?p=2263566&hilit=tabemann#p2263566>
>
> I reckon there are some USB implementation issues which I may be able to
> help with, based on my USB CDC Serial Stack
> <https://github.com/Serialcomms/Raspberry-Pico-USB-Serial_BETA> for the
> RP2040.
>
> Is anyone able to assist with the Zeptopforth-specific elements of this -
> i.e. modules and the console ?
>
> I can help with the zeptoforth-specific parts. If you can contribute to
the zeptoforth USB CDC implementation, go ahead!

Travis

**Comment by Serialcomms** [comment]
OK - great !

two files for you here, install on your 1.9.0-full (non-USB) build on a Pico RP2040

[usb_constants](https://github.com/Serialcomms/zeptoforth-sc/blob/master/usb_constants.fs)

[usb_core](https://github.com/Serialcomms/zeptoforth-sc/blob/master/usb_core.fs)

Run the following for a 'well-behaved' USB serial port which PuTTY and similar clients can connect to reliably
as get/set line coding is implemented. Note that there is **no console** yet - that's still to be done. 

```
- usb-constants import
- usb-core import
- init-usb
- usb-insert-device
```

Lots of debug as USB starts up and a terminal client connects and disconnects. 

I'd like to have CDC Class and CDC Console as separate modules, but not sure how that works in Forth without C-style header files.

**Comment by tabemann** [comment]
On Mon, Nov 25, 2024 at 2:00 PM serialcomms ***@***.***>
wrote:

> OK - great !
>
> two files for you here, install on your 1.9.0-full (non-USB) build on a
> Pico RP2040
>
> usb_constants
> <https://github.com/Serialcomms/zeptoforth-sc/blob/master/usb_constants.fs>
>
> usb_core
> <https://github.com/Serialcomms/zeptoforth-sc/blob/master/usb_core.fs>
>
> Run the following for a 'well-behaved' USB serial port which PuTTY and
> similar clients can connect to reliably
> as get/set line coding is implemented. Note that there is *no console*
> yet - that's still to be done.
>
> - usb-constants import
> - usb-core import
> - init-usb
> - usb-insert-device
>
> Lots of debug as USB starts up and a terminal client connects and
> disconnects.
>
> I'd like to have CDC Class and CDC Console as separate modules, but not
> sure how that works in Forth without C-style header files.
>
You have obviously put a lot of good work into this already, but I have a
minor point to quibble: it isn't safe to use the standard console words
inside an interrupt handler, because if they happen to block they will wait
until the multitasker wakes them, but because the PendSV interrupt handler
is at the lowest priority they will never be woken.

A simple solution would be to wrap all your code where you interact with
the console for debugging prints with:

[: ( PUT YOUR CODE HERE ) ;] debug

where DEBUG would be:

: debug ( xt -- )
  emit-hook @ { saved-emit-hook }
  emit?-hook @ { saved-emit?-hook }
  flush-console-hook @ { saved-flush-console-hook }
  pause-hook @ { saved-pause-hook }
  ['] internal::serial-emit emit-hook !
  ['] internal::serial-emit? emit?-hook !
  [: ;] flush-console-hook !
  [: ;] pause-hook !
  try
  saved-emit-hook emit-hook !
  saved-emit?-hook emit?-hook !
  saved-flush-console-hook flush-console-hook !
  saved-pause-hook pause-hook !
  ?raise
;

What DEBUG does is save the console output hooks and pause hook, set them
for blocking serial output not reliant on the multitasker with PAUSE and
FLUSH-CONSOLE being no-ops, call the passed in xt and catch any exception
raised, restore the hooks to their previous values, and if an exception was
raised then re-raise it.

While strictly speaking this is only necessary within the interrupt handler
at the moment, this will be useful for when the USB CDC console is up and
working and we want to redirect debug output to serial rather than have it
go to the default console, and ultimately this can be easily modified to
disable debugging output altogether without having to physically remove the
debugging prints (even though we will want to use [IF] ... [THEN] to
selectively compile them away when we are all done).

Keep up the good work -- I will try out your code shortly!

Travis

**Comment by tabemann** [comment]
I added DEBUG and I fixed a number of places where structure field names
were overlapping with one another. I also changed all the numeric prints
that were not already using H.8 to using H.8 because . is not always
exactly safe to use in interrupt handlers because it does not play well in
interrupt handlers if code outside them is using something like <# # #> . I
have created a pull request with my changes.

Travis

**Comment by Serialcomms** [comment]
That's tremendous - thanks again Travis !

Next questions are

1. How to reproduce original USB/CDC problem ( I only have RP2040 available )
2. How to prevent module (e.g. usb-core) failing to load when dependent module (e.g. usb-console ) is not loaded yet ?

Early version of [usb-console](https://github.com/Serialcomms/zeptoforth-sc/blob/master/usb_console.fs) (not tested) here

s.c.

**Comment by tabemann** [comment]
On Tue, Nov 26, 2024 at 12:09 PM serialcomms ***@***.***>
wrote:

> That's tremendous - thanks again Travis !
>
> Next questions are
>
>    1. How to reproduce original USB/CDC problem ( I only have RP2040
>    available )
>    2. How to prevent module (e.g. usb-core) failing to load when
>    dependent module (e.g. usb-console ) is not loaded yet ?
>
> Early version of usb-console
> <https://github.com/Serialcomms/zeptoforth-sc/blob/master/usb_console.fs>
> (not tested) here
>
> s.c.
>
> I tried the code with my changes (which I had gotten working on the
RP2040), updated for the RP2350 with zeptocom.js modified to elicit the
original problem (i.e. use a 65535-byte rather than 16-byte buffer), and I
immediately ran into the same problem of USB communications hanging again.
(It worked fine on the RP2350 with picocom.)

Travis

**Comment by Serialcomms** [comment]
Hello again Travis,

Great to see that you got the console working 😄 

I got a working com port, but no console - probably my bad - any instructions ?

Is there a non-debug version available ?

Also want to share these updated words with you for you to include if they help.

[EP1 Handlers](https://github.com/Serialcomms/zeptoforth-sc/blob/master/EP1_HANDLERS.fs)
[Console key & emit](https://github.com/Serialcomms/zeptoforth-sc/blob/master/console-emit-key.fs)

They're adapted from my C stack and are known to work on RP2040.

There were some subtle timing issues which were only exposed via testing.

s.c.

**Comment by tabemann** [comment]
On Wed, Nov 27, 2024, 09:56 serialcomms ***@***.***> wrote:

> Hello again Travis,
>
> Great to see that you got the console working 😄
>
> I got a working com port, but no console - probably my bad - any
> instructions ?
>
> Is there a non-debug version available ?
>
> Also want to share these updated words with you for you to include if they
> help.
>
> EP1 Handlers
> <https://github.com/Serialcomms/zeptoforth-sc/blob/master/EP1_HANDLERS.fs>
> Console key & emit
> <https://github.com/Serialcomms/zeptoforth-sc/blob/master/console-emit-key.fs>
>
> They're adapted from my C stack and are known to work on RP2040.
>
> There were some subtle timing issues which were only exposed via testing.
>
> s.c.
>

In the code as it stands, after connecting a terminal emulator to the
/dev/ttyACMx or, on Windows, COMx device, where x is the number
corresponding to the USB CDC device, you then on your terminal emulator
hooked up via serial to the RP2040 or RP2350 board execute
USB-CONSOLE::USB-CONSOLE . This will activate the USB serial console. Note
that this will need to be changed so it works on boot and with multiple
connections and disconnections.

I also turned off debugging prints in my version of the code. They can be
controlled by setting the constant DEBUG? in USB-CONSTANTS .

Travis

>

**Comment by tabemann** [comment]
Serialcomms,

I now have it working so it automatically starts up the USB CDC console on
bootup after loading usb_constants.fs, usb_core.fs, and usb_console.fs and
then rebooting. I have now started a new pull request with my latest code
so you can merge it in.

I had to wrangle with a troublesome issue on the RP2350 where often the
connection would hang on bootup, but that is now fixed.

However, there is still an issue I noticed where it takes about ten seconds
after rebooting for the USB CDC console to become available to connect to,
and I haven't been able to figure out a solution to that; I had noticed a
very similar issue in the past with the old RP2040/RP2350 USB CDC console
which I had fixed by making ACK packets being sent back in response to all
setup packets, even invalid or unrecognized ones (Linux sends descriptor
request packets that are not recognized), but my old solution to that
problem does not fix this problem for some reason.

Travis

>

**Reply by Serialcomms** [reply]
OK - that's great news and very happy to see it working on RP2350 

10 second delay and unrecognised descriptors may be due to missing **usb-receive-data-packet** below.

( also remove from handle-usb-device-configured )

```

  \ Set USB device configuration 1
  : usb-set-device-configuration ( -- )
    [:

      [ debug? ] [if]
        ." USB Set Device Configuration, Initialising Function Endpoints ... " cr
      [then]
      
      init-usb-function-endpoints
      init-line-nofification 
      init-cdc-line-coding
      8192 EP0-to-Host next-pid !
      EP3-to-Host 10 line-state-notification usb-start-transfer-to-host 

      \ Let host know EP1-to-Pico is working and prepare to receive first data packet
       EP1-to-Pico 64 usb-receive-data-packet

      true prepare-usb-device-configured? !
      usb-ack-out-request
      
    ;] debug
  ;`

```
There's also some confusion over queue-busy? and queue-long? in usb-key which I'm looking at now. 

[C version here](https://github.com/Serialcomms/zeptoforth-sc/blob/master/console-emit-key.fs)

Could do with a new rx-free function for the ring buffer functions - I confused rx-count for that in the above ( now fixed ) 

[usb-cdc-buffers](https://github.com/Serialcomms/zeptoforth-sc/blob/master/usb_cdc_buffers.fs)

All of the ring buffer words can go into a separate module and be loaded between usb-constants and usb-core.

**Comment by tabemann** [comment]
On Thu, Nov 28, 2024 at 4:45 AM serialcomms ***@***.***>
wrote:

> OK - that's great news and very happy to see it working on RP2350
>
> 10 second delay and unrecognised descriptors may be due to missing
> *usb-receive-data-packet* below.
>
> ( also remove from handle-usb-device-configured )
>
>
>   \ Set USB device configuration 1
>   : usb-set-device-configuration ( -- )
>     [:
>
>       [ debug? ] [if]
>         ." USB Set Device Configuration, Initialising Function Endpoints ... " cr
>       [then]
>
>       init-usb-function-endpoints
>       init-line-nofification
>       init-cdc-line-coding
>       8192 EP0-to-Host next-pid !
>       EP3-to-Host 10 line-state-notification usb-start-transfer-to-host
>
>       \ Let host know EP is working and prepare to receive first data packet
>        EP1-to-Pico 64 usb-receive-data-packet
>
>       true prepare-usb-device-configured? !
>       usb-ack-out-request
>
>     ;] debug
>   ;`
>
>
> There's also some confusion over queue-busy? and queue-long? in usb-key
> which I'm looking at now.
>
> C version here
> <https://github.com/Serialcomms/zeptoforth-sc/blob/master/console-emit-key.fs>
>
> Could do with a new rx-free function for the ring buffer functions - I
> confused rx-count for that in the above ( now fixed )
>
> There is no need for QUEUE-BUSY? and QUEUE-LONG? as these can be gotten by
comparing the result from RX-COUNT . As for BUSY? , I am using it to keep
track of when the transactions between the host and the Pico are active.
Also, I moved the call to USB-RECEIVE-DATA-PACKET to
USB::USB-INTERNAL::HANDLE-USB-DEVICE-CONFIGURED . (I am pretty sure the
last is not the source of the problem myself.)

Travis

**Reply by Serialcomms** [reply]
queue-busy? was used in the C stack and can probably be ignored here.
queue-long? is a flag used for USB flow control and is probably more reliable than checking RX-count which
is volatile.

**Comment by tabemann** [comment]
On Thu, Nov 28, 2024 at 9:23 AM serialcomms ***@***.***>
wrote:

> queue-busy? was used in the C stack and can probably be ignored here.
> queue-long? is a flag used for USB flow control and is probably more
> reliable than checking RX-count which
> is volatile.
>
> RX-COUNT is only changed by HANDLE-EP1-TO-PICO and by USB-KEY ;  in the
latter case it checks EP1-TO-PICO BUSY? @ to ensure that HANDLE-EP1-TO-PICO
isn't changing the value returned by RX-COUNT (HANDLE-EP1-TO-PICO would
start a new RX transaction if there were room in the RX circular buffer
anyways).

Travis

**Comment by Serialcomms** [comment]
That's great progress you're making there Travis.

Can you let me know what requests, request-types and recipients you're sending stalls to please ?

Also - would you be able to try [lurk101/pshell multicore for Pico](https://github.com/Serialcomms/zeptoforth-sc/blob/master/pshell_pico_multicore.uf2) on a spare Pico using the same Linux host please ?

Would like to know if it has the same startup delay.

https://github.com/lurk101/pshell

**Comment by Serialcomms** [comment]
Host to Pico flow control is working well !

Can paste a large file into PuTTY window with no lockups, data truncation or other issues 

Does [fast life](https://github.com/tabemann/zeptoforth/blob/master/test/common/fast_life.fs) from your test suite run on this build ?

**Comment by tabemann** [comment]
I am responding to stalls with all setup packets which are not recognized,
which in particular is USB-SETUP SETUP-DESCRIPTOR-TYPE $06.

One thing I should note is that I am having lockup problems with large
uploads on the RP2350, something I will have to solve before I can declare
this done.

Fast life should work fine, even though I haven't tried it in a while. Note
that you will need the proper display for it.

I will try out pshell at some point, but I would like to get the lockup
problems I am seeing solved first.

Travis

Oneh iMessage ID:
***@***.***>

>

**Reply by Serialcomms** [reply]
I’ve ordered a couple of RP2350 Pico 2 boards. Will let you know when they arrive and see if I can help with lockup problems.

**Comment by tabemann** [comment]
On Fri, Nov 29, 2024 at 11:19 AM serialcomms ***@***.***>
wrote:

> I’ve ordered a couple of RP2350 Pico 2 boards. Will let you know when they
> arrive and see if I can help with lockup problems.
>
> I think I have solved the RP2350 lockup problems. The key element was to
precisely time each TX transaction with a hardware timer alarm. Also, in
recent versions of this code I had accidentally introduced a race condition
into RX, which was also causing lockups, which I have since resolved.

Travis

**Comment by tabemann** [comment]
On Fri, Nov 29, 2024 at 6:19 PM Travis Bemann ***@***.***> wrote:

> On Fri, Nov 29, 2024 at 11:19 AM serialcomms ***@***.***>
> wrote:
>
>> I’ve ordered a couple of RP2350 Pico 2 boards. Will let you know when
>> they arrive and see if I can help with lockup problems.
>>
>> I think I have solved the RP2350 lockup problems. The key element was to
> precisely time each TX transaction with a hardware timer alarm. Also, in
> recent versions of this code I had accidentally introduced a race condition
> into RX, which was also causing lockups, which I have since resolved.
>

I tried picoshell and I didn't notice any major delays on bootup, BTW.

Travis

**Reply by Serialcomms** [reply]
That's good to know thanks Travis.

Hope you don't mind, but I've completely re-done the usb console.

Some issues remaining for that

- How to do 'wait for capacity to host' correctly without interfering with USB interrupts ( which will make capacity available )

- How to ensure that USB interrupts are always serviced, regardless of what else is going on, including console waiting or activity

- How to get Core 1 to always service USB interrupts ( to prove we can do multicore like pshell demo )

**Comment by tabemann** [comment]
On Sat, Nov 30, 2024, 12:52 serialcomms ***@***.***> wrote:

> That's good to know thanks Travis.
>
> Hope you don't mind, but I've completely re-done the usb console.
>
> Some issues remaining for that
>
>    -
>
>    How to do 'wait for capacity to host' correctly without interfering
>    with USB interrupts ( which will make capacity available )
>    -
>
>    How to ensure that USB interrupts are always serviced, regardless of
>    what else is going on, including console waiting or activity
>    -
>
>    How to get Core 1 to always service USB interrupts ( to prove we can
>    do multicore like pshell demo )
>
> The thing is that getting it to work truly properly (i.e. never halting
even with repeated very large uploads with zeptocom.js with a very large
buffer size) on the RP2350 requires doing things *very* carefully, and I
had *finally* gotten it to work, so I don't want a rewrite which could
re-break things.

As for how to ensure RX gets restarted even if RX is blocked by the RX
queue being too full, that is accomplished by having USB-KEY restart RX.

As for how to ensure USB interrupts are serviced, that simply is a matter
of never leaving interrupts disabled for long.

As for multicore, I have already taken that into account. There is no need
to service USB interrupts on both cores, and my code uses core-locks (i.e.
special locks that only allow a single core to enter a section of code),
combined with critical sections, to prevent data loss when both cores call
USB-EMIT or USB-KEY simultaneously.

Travis

>

**Reply by Serialcomms** [reply]
No worries and thanks again for info Travis.
Still waiting on Pico 2 delivery.
Presume all still good with RP2040 and 2350 ?

**Comment by Andy2No** [comment]
> I've started to look at the USB/CDC issue reported on the [Raspberry Pico forum](https://forums.raspberrypi.com/viewtopic.php?p=2263566&hilit=tabemann#p2263566)
> 
> I reckon there are some USB implementation issues which I may be able to help with, based on my [USB CDC Serial Stack](https://github.com/Serialcomms/Raspberry-Pico-USB-Serial_BETA) for the RP2040.
> 
> Is anyone able to assist with the Zeptoforth-specific elements of this - i.e. modules and the console ?

Thanks for looking into this.  Would it be appropriate to add a Discussions tab to your repo?

Would it be possible to get your repo working with Earle Philhower's arduino-pico core, for the Arduino IDE? : https://github.com/earlephilhower/arduino-pico 

I probably couldn't help much at all, other than with some beta testing, but I use the RP2040/RP2350 with arduino-pico.  I appreciate that it might be too much work to make your repo work in either setting, at the moment.

**Comment by tabemann** [comment]
On Mon, Dec 2, 2024 at 1:25 AM serialcomms ***@***.***> wrote:

> No worries and thanks again for info Travis.
> Still waiting on Pico 2 delivery.
> Presume all still good with RP2040 and 2350 ?
>
The code I had hacked together works perfectly on both the RP2040 and
RP2350. It could probably benefit from a bit of cleanup, e.g.  separating
out the queues, but even then there are key aspects that need to be taken
into account, e.g. using MOVE to mass-copy data for transmission rather
than repeatedly calling READ-RX to speed things up to decrease interrupt
latency. (Actually, one optimization I could add is using MOVE to directly
copy data into the DPRAM rather than copying it into an intermediate
buffer.)

Travis

**Reply by Serialcomms** [reply]
OK - great

here's the direct dpram copy for you.

```
  : start-ep1-ring-transfer-to-host ( -- )

    tx-count 64 min { tx-bytes }   
    tx-bytes if
    tx-bytes 0 do
    \ de-queue bytes from ring buffer to dpram 
      read-tx EP1-to-Host dpram-address @ i + c! 
    loop

      EP1-to-Host tx-bytes usb-send-data-packet

    then 
  ;

```

**Comment by tabemann** [comment]
On Mon, Dec 2, 2024 at 2:37 AM Andy2No ***@***.***> wrote:

> I've started to look at the USB/CDC issue reported on the Raspberry Pico
> forum
> <https://forums.raspberrypi.com/viewtopic.php?p=2263566&hilit=tabemann#p2263566>
>
> I reckon there are some USB implementation issues which I may be able to
> help with, based on my USB CDC Serial Stack
> <https://github.com/Serialcomms/Raspberry-Pico-USB-Serial_BETA> for the
> RP2040.
>
> Is anyone able to assist with the Zeptoforth-specific elements of this -
> i.e. modules and the console ?
>
> Thanks for looking into this. Would it be appropriate to add a Discussions
> tab to your repo?
>
> Would it be possible to get your repo working with Earle Philhower's
> arduino-pico core, for the Arduino IDE? :
> https://github.com/earlephilhower/arduino-pico
>
> I probably couldn't help much at all, other than with some beta testing,
> but I use the RP2040/RP2350 with arduino-pico. I appreciate that it might
> be too much work to make your repo work in either setting, at the moment.
>
zeptoforth is specifically not Arduino-based, and for that matter contains
no C or C++ code. It will never use arduino-pico because that would
essentially require rewriting it from scratch, for very little practical
benefit.

The recommended IDE environment for zeptoforth is zeptocom.js, which
operates inside one's web browser (it requires a Chromium-based browser
such as Chrome because it relies on Web Serial -- it does not work with
Firefox or mobile browsers); zeptocom.js is online at
https://tabemann.github.io/zeptocomjs/zeptocom.html but if you don't feel
comfortable with a web application accessing your serial ports (despite the
security measures Chrome and like impose) you can git clone the code from
https://github.com/tabemann/zeptocomjs.git, examine it, then run it locally
using python3 -m http.server.

Travis

**Comment by tabemann** [comment]
On Mon, Dec 2, 2024 at 9:04 AM serialcomms ***@***.***> wrote:

> OK - great
>
> here's the direct dpram copy for you.
>
>   : start-ep1-ring-transfer-to-host ( -- )
>
>     tx-count 64 min { tx-bytes }
>     tx-bytes if
>     tx-bytes 0 do
>     \ de-queue bytes from ring buffer to dpram
>       read-tx EP1-to-Host dpram-address @ i + c!
>     loop
>
>       EP1-to-Host tx-bytes usb-send-data-packet
>
>     then
>   ;
>
> That still involves repeatedly calling READ-TX in a DO LOOP, which will be
slow in practice (DO LOOPs are slow, and repeated READ-TX is slower than
mass MOVEs followed by updating TX-READ-INDEX by hand.

What I was thinking of was:

    \ Handle an alarm for sending data
    defer do-tx ( -- )
    :noname ( -- )

      [ rp2040? ] [if]
        tx-alarm timer::clear-alarm-int
        timer::us-counter-lsb
      [then]
      [ rp2350? ] [if]
        tx-alarm timer1::clear-alarm-int
        timer1::us-counter-lsb
      [then]
      { start-us }

      EP1-to-Host busy? @ not if

        tx-write-index @ { write-index }
        tx-read-index @ { read-index }

        write-index read-index <> if
          EP1-to-Host dpram-address @ { dpram }
          0 { total-bytes }

          tx-buffer read-index + { start-addr }
          write-index read-index > if
            write-index read-index - { bytes }
            bytes 64 min to total-bytes
            start-addr dpram total-bytes move
          else
            tx-buffer-size read-index - { first-bytes }
            first-bytes 64 < if
              start-addr dpram first-bytes move
              write-index first-bytes + 64 min first-bytes - { second-bytes
}
              tx-buffer dpram first-bytes + second-bytes move
              first-bytes second-bytes + to total-bytes
            else
              start-addr dpram 64 move
              64 to total-bytes
            then
          then
        then

        EP1-to-Host total-bytes usb-send-data-packet
        total-bytes read-index + $FF and tx-read-index !
      then

      [ rp2040? ] [if]
        start-us tx-interval-us + ['] do-tx tx-alarm timer::set-alarm
      [then]
      [ rp2350? ] [if]
        start-us tx-interval-us + ['] do-tx tx-alarm timer1::set-alarm
      [then]

    ; is do-tx

Travis

**Comment by Serialcomms** [comment]
Before you do that - could we have a look at the ring buffers first please ?

I suspect they're not working as expected in a multicore environment.

Every ring function is potentially affected if read and write are updated simultaneously.

What do you think ?

The Raspberry SDK queues are protected by spinlocks

[Pico SDK Queue .c](https://github.com/raspberrypi/pico-sdk/blob/master/src/common/pico_util/queue.c)

[Pico SDK Queue .h](https://github.com/raspberrypi/pico-sdk/blob/master/src/common/pico_util/include/pico/util/queue.h)

s.c.

**Reply by Serialcomms** [reply]
Hello again Travis,

I've put a new version here for upload and lockup testing only.

[Mini console test ](https://github.com/Serialcomms/zeptoforth-usb-mini-test/tree/main)

It works with your 1.9.1 release on RP2040 and RP2350

There is still a Pico-to-Host buffer issue that I haven't figured out yet - increasing size to 4096 is a workaround for now.

Uploads work well though.

**Comment by tabemann** [comment]
On Sun, Dec 8, 2024 at 4:38 AM serialcomms ***@***.***> wrote:

> Hello again Travis,
>
> I've put a new version here for upload and lockup testing only.
>
> Mini console test
> <https://github.com/Serialcomms/zeptoforth-usb-mini-test/tree/main>
>
> It works with your 1.9.1 release on RP2040 and RP2350
>
> There is still a Pico-to-Host buffer issue that I haven't figured out yet
> - increasing size to 4096 is a workaround for now.
>
> Uploads work well though.
>
> I tried this code (after modifying it to fix a missing END-MODULE
in usb_2_cdc_buffers.fs) and it didn't work (it locked up almost
immediately) when I tried to upload extra/common/zeptoed_all.fs to an
RP2350 board with zeptocom.js with a buffer size (in zeptocom.js) of 65535,
something that I think we should need to be able to do many times in
succession without any issues before we can declare victory. I think the
code I tuned worked pretty well (while did lock up when uploading the
CYW43439 driver and zeptoIP, which are very large, to my Raspberry Pi Pico
2 W, it still was more reliable than anything else I have tested with), and
I think we should use it as a basis for further work rather than starting
over and essentially working from scratch.

Travis

**Comment by Serialcomms** [comment]
Thanks again Travis,

new version on same URL as above.

Now works with default Zeptocom.js ( no buffer changes made )

Try drag-and-drop large files to Coolterm for even faster uploads - 2-3 seconds for your forth-life-test.fs

**Reply by Serialcomms** [reply]
Hello all,

Latest version available on [new github repo](https://github.com/Serialcomms/zeptoforth-usb-mini-test-v2) here.

Features updated ring buffers for improved performance.

Previous version unchanged.

Fast Control-C reboot.

**Reply by Andy2No** [reply]
> Latest version available on [new github repo](https://github.com/Serialcomms/zeptoforth-usb-mini-test-v2) here.

What's the procedure for trying this, e.g. with TeraTerm?  As I'm new to Zeptoforth, it isn't clear to me what to do with it.
 
> Fast Control-C reboot.

Is it possible to do a warm start, back to the REPL, instead of just starting over on a Control-C?

**Reply by Serialcomms** [reply]
@Andy2No 

Steps are

Get zeptoforth.js working with a UART connection to Pico  /Pico 2

Flash Pico with .uf2 ( non USB version ) from Travis's repo

Upload 4 files in numerical order and reboot between each step.

You should get a working USB Serial port with a Zeptoforth prompt.

Use any known working client to access, but not Linux minicom or picocom for now thanks.

Minicom works but does not like reboot or 'surprise' USB removal.

**Comment by tabemann** [comment]
On Tue, Dec 10, 2024 at 2:03 PM serialcomms ***@***.***>
wrote:

> Hello all,
>
> Latest version available on new github repo
> <https://github.com/Serialcomms/zeptoforth-usb-mini-test-v2> here.
>
> Features updated ring buffers for improved performance.
>
> Previous version unchanged.
>
> Fast Control-C reboot.
>
When I tried the latest version (in your new repo) with picocom under
Debian GNU/Linux trixie/sid it would hang after issuing
USB-CONSOLE::USB-CONSOLE from the serial console aside from that it would
display its message saying it was starting the USB console on the serial
console and that it would respond to control-C on the USB console,
rebooting when it was pressed.

Travis

**Comment by Serialcomms** [comment]
Agreed that there are still issues with Linux clients / minicom - possible related to startup / reboot.

Potential workaround in v2 directory, but not confident this is proper fix and only tested with RP Zero as host.

Will build dedicated h/w Linux machine to investigate further, but not until early January. 

Still confident that it works with other clients though.

I will be away with limited access now until early next year, so little or no dev input or engagement from me.
